### PR TITLE
Fixed #9596 -- Added date transform for DateTimeField

### PR DESCRIPTION
**Pull request by jdufresne**
This change allows querying `DateTimeField`s by the date portion. Such as:

```
Articles.objects.filter(published_at__date=datetime.date.today())
```

Some concerns and questions:

1. I do not use Oracle DB. The Oracle implementation was created by reading Oracle documentation and other web searches. It has not been tested, nor is it known to be correct. Feedback from Oracle developers would be appreciated. All other built-in backends were tested.
2. Should I add a `TimeTransform` while I'm at it? To allow users to query on the time portion of a `DateTimeField`. If others think it is a good idea, I wouldn't mind adding it.
3. In DB backend `operations.py`, when converting a column to a timezone, should the timezone be returned as a format parameter to the SQL string (as in MySQL and Postgres) or should the timezone be embedded in the returned SQL string (as in Oracle)?

Thanks. All feedback appreciated.

### CODE REVIEW

**Comment by charettes** [comment]
This would also require a test with `USE_TZ=True`.

**Comment by aaugustin** [comment]
2. `TimeTransform` doesn't sound useful.

3. Use a parameter unless the backend doesn't support it.

File: django/db/backends/sqlite3/operations.py [hunk 1]
```diff
@@ -70,21 +70,23 @@ def date_trunc_sql(self, lookup_type, field_name):
         # cause a collision with a field name).
         return "django_date_trunc('%s', %s)" % (lookup_type.lower(), field_name)
 
+    def _require_pytz(self):
+        if settings.USE_TZ and pytz is None:
+            raise ImproperlyConfigured("This query requires pytz, but it isn't installed.")
+
+    def datetime_cast_date_sql(self, field_name, tzname):
+        self._require_pytz()
+        return "django_datetime_date(%s, %%s)" % field_name, [tzname]
```
Comment (aaugustin) [inline django/db/backends/sqlite3/operations.py]: Can you call this function `django_datetime_cast_date` for consistency? Everything else has cast_date / extract / trunc.

File: docs/releases/1.9.txt [hunk 2]
```diff
@@ -142,6 +142,9 @@ Models
   managers created by ``ForeignKey``, ``GenericForeignKey``, and
   ``ManyToManyField``.
 
+* Added ``date`` lookup to :class:`~django.db.models.DateTimeField`. Can now
+  query a ``DateTimeField`` by only the date portion.
```
Comment (aaugustin) [inline docs/releases/1.9.txt]: s/. Can now query/, which allows querying/

Comment (aaugustin) [inline docs/releases/1.9.txt]: Or anything else that makes a full sentence :-)

**Comment by aaugustin** [comment]
Great work!

Adding the test for `USE_TZ = True` in the `timezones` tests may be the easiest, as there's already some infrastructure you need: time zone definitions and various datetimes values. To prove something, the datetime used in the test should have a different date in UTC and in the current time zone.

**Comment by jdufresne** [comment]
> This would also require a test with USE_TZ=True.

Added.

> Can you call this function django_datetime_cast_date for consistency? Everything else has cast_date / extract / trunc.

Done.

> s/. Can now query/, which allows querying/

Done.

> Adding the test for USE_TZ = True in the timezones tests may be the easiest

I added a `USE_TZ` test to the model_fields tests. Let me know if you think this is enough or you would still prefer to see more tests added to the timezones tests.

---

> TimeTransform doesn't sound useful.

Just from my own experience and application.

General reporting UIs give users tools to make complex queries across a wide range of data. As an example, imagine an application where users enter and record events--an incident log. Using a general reporting tool a user may be interested to study events that happen at night vs during the day. The general reporting tool could allow the user to query by a time. Behind the scenes, the actual query could be implemented using a `TimeTransform` lookup.

I'm happy to add transform, but if people think it is of little value, no problem.

**Comment by akumria** [comment]
The merge conflict here is the release notes; however if you run this patch with warnings on, you will see:

```
test_lookup_date_with_use_tz (model_fields.tests.DateTimeFieldTests) ... /django/django/db/models/fields/__init__.py:1471: RuntimeWarning: DateTimeField DateTimeModel.dt received a naive datetime (2014-03-12 21:22:23.240012) while time zone support is active.

/django/django/db/models/fields/__init__.py:1471: RuntimeWarning: DateTimeField DateTimeModel.dt received a naive datetime (2014-03-11 21:22:23.240011) while time zone support is active.

```
which led us to look more closely (to fix those errors, tzinfo should be passed).

After some discussion with @mjtamlyn we feel the second test may need additional work; it should:

* Pick a time in Vancouver which will result in a different day in UTC (sometime in the evening probably)

* Pick a time in Vancouver which will result in the same day in UTC

* Run the query but set the TZ is UTC, and see only a single result

* Run the query when the TZ is Vancouver, and see both results

You can do this in a single test, using a context manager. I may have a go as I am at a DjangoSprint, but I am recording our findings here in case I run our of time.

File: django/db/models/fields/__init__.py [hunk 3]
```diff
@@ -1489,6 +1489,22 @@ def formfield(self, **kwargs):
         return super(DateTimeField, self).formfield(**defaults)
 
 
+@DateTimeField.register_lookup
```
Comment (akaariai) [inline django/db/models/fields/__init__.py]: Maybe this should be in another file, this file is large enough already.

**Comment by akaariai** [comment]
I'd also like to see an aggregation test case. One of the most common use cases for this is going to be aggregating values per day.

**Comment by jdufresne** [comment]
Thanks for the feedback. I will fix these issues.

> however if you run this patch with warnings on, you will see:

Sorry, but how does one run tests with warnings on? When I run the tests I do not see the warning listed. When I run tests with verbosity 2 I also do not see these warnings. I would like to know how to properly run tests in this way so I can continue debugging.

**Comment by mjtamlyn** [comment]
Run `python -Wall runtests.py ...`. This will print the runtime warnings as above.

**Comment by jdufresne** [comment]
> I'd also like to see an aggregation test case. One of the most common use cases for this is going to be aggregating values per day.

If I understand you correctly you want to see a test that runs a query that groups by the datetime's date for an aggregate result. Something like:

```
DateTimeModel.objects.values('dt__date').aggregate(...)
```

Is this correct? If so, it appears the ORM does not support lookup fields in a values clause. I receive the following error with this query:

```
Traceback (most recent call last):
  File "/home/jon/devel/django/django/test/utils.py", line 182, in inner
    return test_func(*args, **kwargs)
  File "/home/jon/devel/django/tests/model_fields/tests.py", line 297, in test_lookup_date_with_use_tz
    qs = DateTimeModel.objects.values('dt__date').annotate(date_count=Count('id'))
  File "/home/jon/devel/django/django/db/models/manager.py", line 127, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/jon/devel/django/django/db/models/query.py", line 703, in values
    clone = self._values(*fields)
  File "/home/jon/devel/django/django/db/models/query.py", line 698, in _values
    query.add_fields(field_names, True)
  File "/home/jon/devel/django/django/db/models/sql/query.py", line 1603, in add_fields
    name.split(LOOKUP_SEP), opts, alias, allow_many=allow_m2m)
  File "/home/jon/devel/django/django/db/models/sql/query.py", line 1377, in setup_joins
    names, opts, allow_many, fail_on_missing=True)
  File "/home/jon/devel/django/django/db/models/sql/query.py", line 1345, in names_to_path
    " not permitted." % (names[pos + 1], name))
FieldError: Cannot resolve keyword u'date' into field. Join on 'dt' not permitted.
```

What are you thoughts regarding this?

**Comment by jdufresne** [comment]
> Maybe this should be in another file, this file is large enough already.

I have no problem with this, but what would be a more appropriate location? I didn''t put it inside `django.db.models.lookups` (with the `default_lookups`) to avoid the circular import for `DateTimeField.register_lookup()`. Also, in my opinion it seems reasonable to keep it close to `DateTimeField`. What did you have in mind?

**Comment by jdufresne** [comment]
@akumria I have addressed your test concerns with the most recent version. Cheers.

File: tests/model_fields/tests.py [hunk 4]
```diff
@@ -265,6 +265,38 @@ def test_datetimes_save_completely(self):
...
+        dt2 = datetime.datetime(2014, 3, 11, 21, 22, 23, 240000)
+        t = datetime.time(21, 22, 23, 240000)
+        m = DateTimeModel.objects.create(d=d, dt=dt1, t=t)
+        # Other model with different datetime.
+        DateTimeModel.objects.create(d=d, dt=dt2, t=t)
+        self.assertEqual(m, DateTimeModel.objects.get(dt__date=d))
+
+    @test.override_settings(USE_TZ=True, TIME_ZONE='America/Vancouver')
+    def test_lookup_date_with_use_tz(self):
+        d = datetime.date(2014, 3, 12)
```
Comment (akumria) [inline tests/model_fields/tests.py]:
```python
# the following time is equivalent to UTC 2014-03-12 18:34:23.24000
```

File: tests/model_fields/tests.py [hunk 5]
```diff
@@ -265,6 +265,38 @@ def test_datetimes_save_completely(self):
...
+        dt1 = datetime.datetime(
+            2014, 3, 12, 10, 22, 23, 240000,
+            tzinfo=timezone.get_current_timezone())
```
Comment (akumria) [inline tests/model_fields/tests.py]:
```python
# the following time is equivalent to UTC 2014-03-13 05:34:23.24000
```

File: tests/model_fields/tests.py [hunk 6]
```diff
@@ -265,6 +265,38 @@ def test_datetimes_save_completely(self):
...
+        dt2 = datetime.datetime(
+            2014, 3, 12, 21, 22, 23, 240000,
+            tzinfo=timezone.get_current_timezone())
+        t = datetime.time(21, 22, 23, 240000)
+        m1 = DateTimeModel.objects.create(d=d, dt=dt1, t=t)
+        m2 = DateTimeModel.objects.create(d=d, dt=dt2, t=t)
```
Comment (akumria) [inline tests/model_fields/tests.py]:
```python
# In Vancouver, we expect both results
```

File: tests/model_fields/tests.py [hunk 7]
```diff
@@ -265,6 +265,38 @@ def test_datetimes_save_completely(self):
...
+        self.assertQuerysetEqual(
+            DateTimeModel.objects.filter(dt__date=d),
+            [repr(m1), repr(m2)],
+            ordered=False)
```
Comment (akumria) [inline tests/model_fields/tests.py]:
```python
# but in UTC, the __date only matches one of them
```

**Comment by akumria** [comment]
@mjtamlyn lgtm. I'm unclear why djangoci has flagged this as failing, it works locally.

@jdufresne I added in the extra comments, as they helped me understand the rationale of the testcase (especially handy in the future when people may not have the PR context).

**Comment by timgraham** [comment]
This should likely also be documented somewhere in ref and/or topics.

File: docs/releases/1.9.txt [hunk 8]
```diff
@@ -167,6 +167,9 @@ Models
 * Added a system check to prevent defining both ``Meta.ordering`` and
   ``order_with_respect_to`` on the same model.
 
+* Added ``date`` lookup to :class:`~django.db.models.DateTimeField`, which
+  allows querying a ``DateTimeField`` by only the date portion.
```
Comment (timgraham) [inline docs/releases/1.9.txt]: We should add something like "To use this feature, third-party database backends may need to implement a `datetime_cast_date_sql()` method." (might be better in a "Database backend API" section like we have in the 1.8 release notes).

File: tests/model_fields/tests.py [hunk 9]
```diff
@@ -236,7 +236,7 @@ def test_abstract_model_pending_lookups(self):
         )
 
 
-class DateTimeFieldTests(unittest.TestCase):
+class DateTimeFieldTests(test.TestCase):
```
Comment (timgraham) [inline tests/model_fields/tests.py]: could you fix the other test cases in this file and make it a separate commit? I see at least `BooleanFieldTests` that has database activity but uses `unittest.TestCase`.

**Comment by timgraham** [comment]
buildbot, test on oracle.

**Comment by jdufresne** [comment]
Thanks @timgraham. I made updates to the commits based on your feedback. Additional feedback is welcome.

**Comment by timgraham** [comment]
Did you see the failures on Oracle? http://djangoci.com/view/PR%20Builders/job/pull-requests-oracle/12/

**Comment by jdufresne** [comment]
Thanks @timgraham I will take a look at installing Oracle to debug the issue.

**Comment by aaugustin** [comment]
I had a look at the tests you added since I last reviewed this PR.

They will fail:
- under Windows
- with MySQL when time zone definitions aren't installed.

You should decorate them with:

    @requires_tz_support
    @skipUnlessDBFeature('has_zoneinfo_database')

That's why I suggested adding timezone-related tests in the timezones tests which already have suitable infrastructure to skip tests on some platforms.

**Comment by jdufresne** [comment]
@timgraham I have addressed the test failures. Turns out, they were actually caused by the changes to `BooleanFieldTests` and unrelated to any recent date work.

@aaugustin I have added the decorators and tested the the test is properly skipped on MySQL without zoneinfo. I do not have a Windows machine and so am unable to test that directly.

**Comment by jdufresne** [comment]
Why does the Jenkins CI server occasionally fail with the follow error? Is there something I can do to prevent this? It looks intermittent and unrelated to my changes.

```
> git checkout -f 317d0c2bb0e391223b1cebc0742f420bd95e8d8a
FATAL: Could not checkout null with start point 317d0c2bb0e391223b1cebc0742f420bd95e8d8a
hudson.plugins.git.GitException: Could not checkout null with start point 317d0c2bb0e391223b1cebc0742f420bd95e8d8a
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$8.execute(CliGitAPIImpl.java:1855)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:152)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:145)
	at hudson.remoting.UserRequest.perform(UserRequest.java:121)
	at hudson.remoting.UserRequest.perform(UserRequest.java:49)
	at hudson.remoting.Request$2.run(Request.java:325)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
	at java.util.concurrent.FutureTask.run(FutureTask.java:166)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1146)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:701)
	at ......remote call to ci7(Native Method)
	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1360)
	at hudson.remoting.UserResponse.retrieve(UserRequest.java:221)
	at hudson.remoting.Channel.call(Channel.java:753)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:145)
	at sun.reflect.GeneratedMethodAccessor348.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:622)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:131)
	at com.sun.proxy.$Proxy51.execute(Unknown Source)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1033)
	at hudson.scm.SCM.checkout(SCM.java:484)
	at hudson.model.AbstractProject.checkout(AbstractProject.java:1270)
	at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:609)
	at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:531)
	at hudson.model.Run.execute(Run.java:1741)
	at hudson.matrix.MatrixRun.run(MatrixRun.java:146)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:374)
Caused by: hudson.plugins.git.GitException: Command "git checkout -f 317d0c2bb0e391223b1cebc0742f420bd95e8d8a" returned status code 128:
stdout: 
stderr: fatal: reference is not a tree: 317d0c2bb0e391223b1cebc0742f420bd95e8d8a

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1591)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$500(CliGitAPIImpl.java:86)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$8.execute(CliGitAPIImpl.java:1850)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:152)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:145)
	at hudson.remoting.UserRequest.perform(UserRequest.java:121)
	at hudson.remoting.UserRequest.perform(UserRequest.java:49)
	at hudson.remoting.Request$2.run(Request.java:325)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
	at java.util.concurrent.FutureTask.run(FutureTask.java:166)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1146)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:701)
Recording test results
ERROR: Publisher hudson.tasks.junit.JUnitResultArchiver aborted due to exception
hudson.AbortException: Test reports were found but none of them are new. Did tests run? 
For example, /home/jenkins/workspace/pull-requests-trusty/database/postgres/label/trusty-pr/python/python2.7/tests/TEST-absolute_url_overrides.tests.AbsoluteUrlOverrideTests.xml is 3 hr 12 min old

	at hudson.tasks.junit.TestResult.parse(TestResult.java:178)
	at hudson.tasks.junit.TestResult.parse(TestResult.java:146)
	at hudson.tasks.junit.TestResult.<init>(TestResult.java:122)
	at hudson.tasks.junit.JUnitParser$ParseResultCallable.invoke(JUnitParser.java:119)
	at hudson.tasks.junit.JUnitParser$ParseResultCallable.invoke(JUnitParser.java:93)
	at hudson.FilePath$FileCallableWrapper.call(FilePath.java:2688)
	at hudson.remoting.UserRequest.perform(UserRequest.java:121)
	at hudson.remoting.UserRequest.perform(UserRequest.java:49)
	at hudson.remoting.Request$2.run(Request.java:325)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
	at java.util.concurrent.FutureTask.run(FutureTask.java:166)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1146)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:701)
	at ......remote call to ci7(Native Method)
	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1360)
	at hudson.remoting.UserResponse.retrieve(UserRequest.java:221)
	at hudson.remoting.Channel.call(Channel.java:753)
	at hudson.FilePath.act(FilePath.java:978)
	at hudson.FilePath.act(FilePath.java:967)
	at hudson.tasks.junit.JUnitParser.parseResult(JUnitParser.java:90)
	at hudson.tasks.junit.JUnitResultArchiver.parse(JUnitResultArchiver.java:120)
	at hudson.tasks.junit.JUnitResultArchiver.perform(JUnitResultArchiver.java:137)
	at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:74)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:761)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:721)
	at hudson.model.Build$BuildExecution.post2(Build.java:183)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:670)
	at hudson.model.Run.execute(Run.java:1766)
	at hudson.matrix.MatrixRun.run(MatrixRun.java:146)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:374)
Finished: FAILURE
```

**Comment by timgraham** [comment]
Nothing you can do unless you want to fix Jenkins bugs, I think. See https://code.djangoproject.com/wiki/Jenkins#Troubleshooting. Not a big deal though -- if most of the matrix passes, your change is likely fine.

**Comment by jdufresne** [comment]
From @akaariai 

> Maybe this should be in another file, this file is large enough already.

I have thought about this some more.

One idea I had would be to split the `django/db/models/fields/__init__.py` file up. I was thinking one of the following:

* `__init__.py` top level file used as an interface for importing outside the module.
* `base.py` effectively a copy of the current `__init__.py`, but without the date fields
* `dates.py` all the date and time fields plus the date and time transforms (if the changes in PR #4393  get fleshed out, this would contain all those transforms as well)

Alternatively, we could separate the fields from the lookups/transforms. Something like:

* `__init__.py` top level file used as an interface for importing outside the module.
* `base.py` effectively a copy of the current `__init__.py`, but without the new lookups
* `lookups.py` all the lookups and transforms, register them on the fields (again, if the changes in PR #4393  get fleshed out, this would contain all those transforms as well)

I would only to go ahead with this if others agree and think this should be a part of these changes. If I do go ahead, the renaming could go as a separate commit before the existing ones.

Thoughts on if this is an improvement and addresses your concerns?

**Comment by mjtamlyn** [comment]
I've got some wider ideas here as well to do with removing junk transforms from fields which make no sense to have them. We perhaps want to join forces somewhat and do several patches here.

- The first change is this one, adding `__date`.
- The second change is moving things around. We want a `__init__.py`, a `base.py`, a `text.py` (for fields relating to textual data, i.e. where `__contains` is meaningful), possibly a `numbers.py`, a `dates.py`, a `special.py` for more specialized fieds if needed. Lookups specific to each section can live after the definition of the fields, as we do in `contrib.postgres`.
- The third change is to convert all the date lookups into transforms and remove them from all other fields. At the same time, an addition to `contrib.postgres` needs to be made with `__decade`, `__century` etc.
- The fourth change is to do a general cleanup of the built in lookups, removing lookups which cannot work from fields and grouping them sensibly. For example, I don't feel that `__startswith` should work on a date field for example. However, we should add a `__text` transform which applies to all fields if this is possible. On postgres this is simply `::text`.

**Comment by akumria** [comment]
It sounds like 4 separate pull requests would be easier, rather than
piling additional work onto this one.

A


On 02/04/15 14:49, Marc Tamlyn wrote:
>
> I've got some wider ideas here as well to do with removing junk
> transforms from fields which make no sense to have them. We perhaps
> want to join forces somewhat and do several patches here.
>
>   * The first change is this one, adding |__date|.
>   * The second change is moving things around. We want a
>     |__init__.py|, a |base.py|, a |text.py| (for fields relating to
>     textual data, i.e. where |__contains| is meaningful), possibly a
>     |numbers.py|, a |dates.py|, a |special.py| for more specialized
>     fieds if needed. Lookups specific to each section can live after
>     the definition of the fields, as we do in |contrib.postgres|.
>   * The third change is to convert all the date lookups into
>     transforms and remove them from all other fields. At the same
>     time, an addition to |contrib.postgres| needs to be made with
>     |__decade|, |__century| etc.
>   * The fourth change is to do a general cleanup of the built in
>     lookups, removing lookups which cannot work from fields and
>     grouping them sensibly. For example, I don't feel that
>     |__startswith| should work on a date field for example. However,
>     we should add a |__text| transform which applies to all fields if
>     this is possible. On postgres this is simply |::text|.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/django/django/pull/4281#issuecomment-88908442>.
>


-- 
“Don’t be sad because it’s over. Smile because it happened.” – Dr. Seuss

**Comment by timgraham** [comment]
buildbot, test on oracle.

**Comment by timgraham** [comment]
confirmed tests pass on Windows

**Comment by timgraham** [comment]
Are there objections to merging this after my comments are addressed?

**Comment by timgraham** [comment]
Still an issue with Oracle, but the tests do pass locally so I guess it's related to a different Oracle version on CI or something.

**Comment by jdufresne** [comment]
Thanks Tim. I have incorporated your feedback into my latest changes. Any additional feedback is welcome.

I will look into installing an Oracle test server to see if I can replicate the test failure.

**Comment by jdufresne** [comment]
> I will look into installing an Oracle test server to see if I can replicate the test failure.

The instructions at <https://code.djangoproject.com/wiki/OracleTestSetup#using-the-oracle-developer-day-pre-installed-vm> seems incomplete or outdated. After following these instructions and trying to account for some differences, I'm still unable to get a working Oracle test environment.

Doing my best to understand what the differences should be and what might or might not work, I'm left with the following below. I'm not sure how to proceed to get a working test environment. Any help getting past this would be appreciated.

```
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.oracle',
        'HOST': '127.0.0.1',
        'PORT': '1521',
        'NAME': 'cdb1',
        'USER': 'system',
        'PASSWORD': 'oracle',
    },
    'other': {
        'ENGINE': 'django.db.backends.oracle',
        'HOST': '127.0.0.1',
        'PORT': '1521',
        'NAME': 'cdb1',
        'USER': 'system',
        'PASSWORD': 'oracle',
    }
}
```

```
Testing against Django installed in '/home/jon/devel/django/django'
Creating test database for alias 'default'...
Failed (ORA-01543: tablespace 'TEST_SYSTEM' already exists
)
Got an error creating the test database: ORA-01543: tablespace 'TEST_SYSTEM' already exists

It appears the test database, test_system, already exists. Type 'yes' to delete it, or 'no' to cancel: yes
Destroying old test database 'default'...
Creating test user...
Failed (ORA-65096: invalid common user or role name
)
Got an error creating the test user: ORA-65096: invalid common user or role name

It appears the test user, test_system, already exists. Type 'yes' to delete it, or 'no' to cancel: yes
Destroying old test user...
Failed (ORA-01918: user 'TEST_SYSTEM' does not exist
)
Got an error recreating the test user: ORA-01918: user 'TEST_SYSTEM' does not exist
```

**Comment by timgraham** [comment]
Can you try with an exact copy of the settings file  (instead of with a different `NAME`)? Also, don't omit the `DATABASES['TEST']` setting.

But the problem is, I couldn't reproduce the test failure with the Developers Day VM, so it might be a difference between that Oracle version and what we have on CI.

**Comment by jdufresne** [comment]
> Can you try with an exact copy of the settings file (instead of with a different NAME)? Also, don't omit the DATABASES['TEST'] setting.

Yes. See the output below, which is why I changed a few things to get farther.

The reason I say the wiki looks outdated is because the wiki is describing an OVA file for Database11g Release 2 and instantclient 11.2, while the current version provided by Oracle's is an OVA with Database 12c Release 1 and instantclient 12.1.

As these are the versions available for download, these are the ones I'm using. I'm not positive the instructions provided in the wiki work exactly for them without any changes. Changing `'NAME`: 'orcl'` to `cdb1` at least allows a connection to be establish.

> But the problem is, I couldn't reproduce the test failure with the Developers Day VM, so it might be a difference between that Oracle version and what we have on CI.

Is CI using 12c Release 1, 11g Release 2, or something else?

Anyway, using the exact settings.py, here is my results:

```
Testing against Django installed in '/home/jon/devel/django/django'
Creating test database for alias 'default'...
Traceback (most recent call last):
  File "./runtests.py", line 434, in <module>
    options.debug_sql)
  File "./runtests.py", line 256, in django_tests
    extra_tests=extra_tests,
  File "/home/jon/devel/django/django/test/runner.py", line 210, in run_tests
    old_config = self.setup_databases()
  File "/home/jon/devel/django/django/test/runner.py", line 166, in setup_databases
    **kwargs
  File "/home/jon/devel/django/django/test/runner.py", line 370, in setup_databases
    serialize=connection.settings_dict.get("TEST", {}).get("SERIALIZE", True),
  File "/home/jon/devel/django/django/db/backends/base/creation.py", line 57, in create_test_db
    self._create_test_db(verbosity, autoclobber, keepdb)
  File "/home/jon/devel/django/django/db/backends/oracle/creation.py", line 17, in _create_test_db
    cursor = self.connection.cursor()
  File "/home/jon/devel/django/django/db/backends/base/base.py", line 164, in cursor
    cursor = self.make_cursor(self._cursor())
  File "/home/jon/devel/django/django/db/backends/base/base.py", line 135, in _cursor
    self.ensure_connection()
  File "/home/jon/devel/django/django/db/backends/base/base.py", line 130, in ensure_connection
    self.connect()
  File "/home/jon/devel/django/django/db/utils.py", line 95, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/jon/devel/django/django/db/backends/base/base.py", line 130, in ensure_connection
    self.connect()
  File "/home/jon/devel/django/django/db/backends/base/base.py", line 119, in connect
    self.connection = self.get_new_connection(conn_params)
  File "/home/jon/devel/django/django/db/backends/oracle/base.py", line 214, in get_new_connection
    return Database.connect(conn_string, **conn_params)
django.db.utils.DatabaseError: ORA-12505: TNS:listener does not currently know of SID given in connect descriptor
```

**Comment by timgraham** [comment]
Oh, I have version 11.2.0.2.0 of the Developers Day VM, so yes, the instructions may need to be updated for version 12. We are using 11.2.0.1.0 on Jenkins.

**Comment by timgraham** [comment]
Hi Jon, you didn't get version 12 of the VM working did you? I've been trying to update the instructions, but I get stuck with "ORA-65096: invalid common user or role name" when running the tests. Using same database settings from the wiki except `NAME: 'cdb1'`.

**Comment by jdufresne** [comment]
> Hi Jon, you didn't get version 12 of the VM working did you? I've been trying to update the instructions, but I get stuck with "ORA-65096: invalid common user or role name" when running the tests. Using same database settings from the wiki except NAME: 'cdb1'.

No I did not. Unfortunately I am very inexperience with Oracle and was having a lot of issues debugging the root cause. I never discovered a successful setup.

I was planning to either try the older Oracle VM (although from what I understand this environment passes tests) or wait for further development on the Django Docker/Vagrant work being discussed on the mailing list.

File: django/db/backends/oracle/operations.py [hunk 10]
```diff
@@ -127,20 +129,18 @@ def _convert_field_to_tz(self, field_name, tzname):
         # on DATE values, even though they actually store the time part.
         return "CAST(%s AS TIMESTAMP)" % result
 
+    def datetime_cast_date_sql(self, field_name, tzname):
+        field_name = self._convert_field_to_tz(field_name, tzname)
+        sql = 'CAST(%s AS DATE)' % field_name
```
Comment (timgraham) [inline django/db/backends/oracle/operations.py]: the tests pass for me on Oracle using `'TRUNC(%s)'`

**Comment by jdufresne** [comment]
> the tests pass for me on Oracle using 'TRUNC(%s)'

 Thanks @timgraham. I made this change in the latest version of the commit.

**Comment by timgraham** [comment]
buildbot, test on oracle.

**Comment by timgraham** [comment]
merged in 44f3ee77166bd5c0e8a4604f2d96015268dce100, thanks!

**Comment by Bakirelived** [comment]
I think i found a 'bug' or at least an issue that can lead to unexpected behavior, when using __date filed lookup and timezones. I'm iterating and processing a queryset, witch has a timestamp that can range multiple days, and I'm processing daily data from start to finish. At the end, i exclude the processed day's data and reiterate.

The problem is when the day changes, I'm using set = set.exclude(timestamp__date=set.first().timestamp.date()), but It doesn't filter out anything(and it should, at least the first, and yes, set it is not empty).

set has one value, with timestamp 2016-08-24 23:00:31.405251+00:00, 
set.first().timestamp.date() has the value of 2016-08-24

The problem here is with the use_tz, as i have it true, and when the filter evaluates, it makes the two values have different dates.

**Comment by timgraham** [comment]
Please ask "is it a bug?" questions on our [our support channels](https://code.djangoproject.com/wiki/TicketClosingReasons/UseSupportChannels) and [create a ticket](https://code.djangoproject.com/newticket) if you get confirmation of that. Thanks.

**Comment by charettes** [comment]
@Bakirelived as datetimes are returned in UTC by the ORM and `__dates uses the correctly active timezone you'll have to use `django.utils.timezone.localtime` on your input.

```
set.exclude(timestamp__date=localtime(set.first().timestamp).date())
```

**Comment by Bakirelived** [comment]
thank you very much, that was exactly the problem. As my timezone is close to UTC, I didn't know that datetimes are always UTC.. always learning.. but I fell that this is one of those warnings that could be in the new __date lookup field documentation, following the current explanation of the use_tz effect
